### ISSUE TITLE
the output of pointsops.knnquery out of index as commented 'may have invalid neighbor'

### ISSUE THREAD

**Issue by guxss3** [LiyaoTang/contrastBoundary#8]
Hi,@LiyaoTang  
Thanks for you excellent work. I met the question as you commented in basic_operators.py that the return 'neighbor_idx' is out of index at 'n=up' and 'i=4'. 
Specifically, I set batchsize = 2 to train s3dis in one gpu for running the segmentation pipeline of CBL . the shape of 'x'('target') is 152706 while the maximum of 'neighbor_idx' is 1048258969. So how do i  to solve the error? Could I just filter out the wrong value in 'neighbor_idx'?
Thanks.

**Comment by LiyaoTang** [comment]
I have left some printing codes there for debug use. Could you check the value of `kr` first? If it's too large (eg. larger than the input points of each example), you may get invalid neighbors or run out of the GPU mem. Besides, did you modify the code anywhere else?

**Comment by guxss3** [comment]
Hi, sorry for reply so later, and for my next tedious balabala.
Firstly, the gpu mem is enough and I just modified the config file so that I can train on one single gpu. Specifically, I set batchsize =2  and 'train_gpu' as '[0]'.
And, I checked the value of 'kr' and debug the code, 'kr' is from 4 to 256, corresponding to the head cfg 'nstride'. I think it is no ok. 
Besides, I add some codes in basic_operators.py to fillter the invalid neighbors . Here it is:
    # print('kr - x', kr, x.shape)
    # print(p_from.shape, p_to.shape)
    # print(o_from, o_to)
    # print('neighbor_idx.shape = ', neighbor_idx.shape)
    # print(neighbor_idx.min(), neighbor_idx.max(), (neighbor_idx == neighbor_idx.max()).int().sum())

    # neighbor_idx[neighbor_idx > p.shape[0]] = p.shape[0]  # collect all 0s if invalid neighbor
    # x = torch.cat([x, torch.zeros([1, self.config.num_classes])])

    neighbor_idx = neighbor_idx.view(-1).long()
#######[DEBUG]    the return neighbor_idx of function knnquery is out of index, I just filter and use nearest neighbor to #######replace it
    if neighbor_idx.max() >= x.shape[0] or neighbor_idx.min() < 0:
        out_of_index = torch.cat((torch.where(neighbor_idx >= x.shape[0])[0],torch.where(neighbor_idx < 0)[0]))
        print('[WARNING] There are %d indexes overflow at stage_n = %s, stage_i = %s \n'%(out_of_index.shape[0],stage_n, stage_i))
        for index in out_of_index:
            i = 1
            while index+i<x.shape[0] or index-i>=0:
                if not (index+i in out_of_index):
                    neighbor_idx[index] = neighbor_idx[index+i]
                    break
                if not (index - i  in out_of_index):
                    neighbor_idx[index] = neighbor_idx[index-i]
                    break
                i+=1
    x = x[neighbor_idx, :].view(p_to.shape[0], kr, x.shape[1]) # (m, kr, ncls)
    x = x.float().mean(-2)  # (m, ncls)
##########################[DEBUG END]

    # x = x.float().sum(-2)  # (m, ncls)
    # cnt = (neighbor_idx < p.shape[0]).float().sum(-1, keepdim=True)  # (m, 1)
    # x /= cnt
After added the codes, the training pipeline could run, but it is so slow becase the while loop in my debug code. Here is a small part of  output log:
[2022-04-15 13:35:31,810 INFO train.py line 354 1668] Epoch: [1/100][1/120] Data 3.098 (3.098) Batch 114.310 (114.310) Remain 381:00:01 Loss [2.5864 0.0299 0.0306 0.0311 0.0473 0.004 ] Accuracy 0.0957.
[WARNING] There are 52201 indexes overflow at stage_n = up, stage_i = 4 

[2022-04-15 13:36:59,105 INFO train.py line 354 1668] Epoch: [1/100][2/120] Data 2.705 (2.902) Batch 87.295 (100.803) Remain 335:57:08 Loss [2.7387 0.0298 0.0312 0.0363 0.0564 0.    ] Accuracy 0.2368.
[WARNING] There are 16899 indexes overflow at stage_n = up, stage_i = 4 

[2022-04-15 13:37:25,856 INFO train.py line 354 1668] Epoch: [1/100][3/120] Data 0.425 (2.076) Batch 26.751 (76.119) Remain 253:39:53 Loss [2.3618 0.0303 0.0305 0.0419 0.057  0.    ] Accuracy 0.2892.
[WARNING] There are 40962 indexes overflow at stage_n = up, stage_i = 4  

I just can not figure out why the function knnquery return lots of invalid neighbors while the repo point-transformer (your backbone model, [https://github.com/POSTECH-CVLab/point-transformer](url)) is normally trained and valided both in s3dis and my own dataset. 

Another thing I want to consult is that the mIoU is obviously lower than that of point-transformer during the first few epochs(about 6), which is expected? 

Best.

**Comment by LiyaoTang** [comment]
Could you provide me with the output of those printing codes when the error occurs? (together with the error message).

Thanks.
